sling/nlp/wikicat/prelim_ranker.py

```diff
@@ -0,0 +1,208 @@
+# Copyright 2017 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http:#www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Task for outputting a preliminary ranking of parses, keeping only the top-k,
+# and computing a signature for each surviving parse.
+# A signature replaces each (PID, QID) span in a parse by $PID=$label(QID),
+# where 'label' is derived using a custom taxonomy.
+
+import sling
+from sling.task.workflow import register_task
+from util import load_kb
+
+class ParseRanker:
+  def init(self, task):
+    self.kb = load_kb(task)
+    self.extractor = sling.api.FactExtractor(self.kb)
+    self.h_name = self.kb['name']
+
+    # Custom type ordering for building a taxonomy.
+    taxonomy_types = [
```

Comment:
I think  you should have way fewer types in the taxonomy and also remove some of the general types like concept and entity. Otherwise will end up with signatures that have very little lexicalization. The taxonomy should only contain the types that you want to generalize over.

Reply:
I have removed four entries for now (including concept, entity). Most of the remaining ones look good "mid-level" types to me that shouldn't be lexicalized. That being said, I will do some more eyeballing and tighten the taxonomy further as needed.